server/src/main/java/io/crate/metadata/doc/DocTableInfo.java

```diff
@@ -674,12 +674,41 @@ private void validateDropColumns(List<DropColumn> dropColumns) {
         }
     }
 
+    /**
+     * Propagates the changes occurred to child columns to the parent columns inner types, e.g.::
+     *   ALTER TABLE ADD COLUMN o['o2']['o3']['x'] INT;
+     *   ==> after adding the column `x`, column `o`, `o['o2']`, and `o['o2']['o3']` also need to append `x` to its inner types.
+     * @param column the column changed
+     * @param type the type that the column changed to (null if dropped)
+     * @param newReferences all references of the current table including the changed references
+     */
+    private void fixParentsInnerTypes(ColumnIdent column,
+                                      @Nullable DataType<?> type,
+                                      Map<ColumnIdent, Reference> newReferences) {
+        var parent = column.getParent();
+        if (parent == null) {
+            return;
+        }
+        Reference parentRef = newReferences.get(parent);
+        if (parentRef == null) {
+            throw new ColumnUnknownException(column, ident);
+        }
+        DataType<?> newParentType = ArrayType.updateLeaf(
+            parentRef.valueType(),
+            leaf -> type == null ?
+                ((ObjectType) leaf).withoutChild(column.leafName()) :
+                ((ObjectType) leaf).withChild(column.leafName(), type)
+        );
+        Reference updatedParent = parentRef.withValueType(newParentType);
+        newReferences.replace(parent, updatedParent);
+        fixParentsInnerTypes(updatedParent.column(), updatedParent.valueType(), newReferences);
```

Comment:
Not sure if it's worth doing, but couldn't this easily avoid the recursion (stack depth) by looping over the parents? (`ColumnIdent` already has a `parents()` iterator)

Reply:
Thank you, updated.